### PR TITLE
Add support for dollar-quoted strings

### LINKED ISSUE #11512: Support `$$` Dollar-quoted String Constants
**Use case**:

> _**From postgresql.org**_
> While the standard syntax for specifying string constants is usually convenient, it can be difficult to understand when the desired string contains many single quotes or backslashes, since each of those must be doubled. To allow more readable queries in such situations, PostgreSQL provides another way, called “dollar quoting”, to write string constants.

replace
``` sql
INSERT INTO a_table (txt) VALUES ('Quotes shouldn''t be doubled')
```
with
``` sql
INSERT INTO a_table (txt) VALUES ($$ Quotes shouldn't be doubled $$)
```

---

replace
``` sql
INSERT INTO a_job_table (job) VALUES (
'INSERT INTO "tab_A"
    SELECT * FROM tab_b
    WHERE col1 = ''a'''
)
```
with
``` sql
INSERT INTO a_job_table (job) VALUES (
$$ INSERT INTO "tab_A"
    SELECT * FROM tab_b
    WHERE col1 = 'a'
$$
)
```

---

replace
```sql
CREATE FUNCTION symmetric_point(point geo_point)
RETURNS geo_point
LANGUAGE JAVASCRIPT
AS 
'function symmetric_point (point, angle) {
      var x = - point[0],
          y = - point[1];
      return ''POINT ('' + x + '', '' + y +'')'';
   }';
```
with
```sql
CREATE FUNCTION symmetric_point(point geo_point)
RETURNS geo_point
LANGUAGE JAVASCRIPT
AS 
$function$
function symmetric_point (point, angle) {
      var x = - point[0],
          y = - point[1];
      return 'POINT (' + x + ', ' + y +')';
   }
$function$;
```

**Feature description**:

> _**From postgresql.org**_
> A dollar-quoted string constant consists of a dollar sign ($), an optional “tag” of zero or more characters, another dollar sign, an arbitrary sequence of characters that makes up the string content, a dollar sign, the same tag that began this dollar quote, and a dollar sign. For example, here are two different ways to specify the string “Dianne's horse” using dollar quoting:

Support `$$` as alternative  for string constants

```sql
$$Dianne's horse$$
$SomeTag$Dianne's horse$SomeTag$
```

also see postgres implementation:
https://www.postgresql.org/docs/13/sql-syntax-lexical.html#SQL-SYNTAX-DOLLAR-QUOTING

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Resolves #11512.

Add support for dollar quoted strings [that cannot handle nested dollar quoting yet](https://github.com/crate/crate/issues/11512#issuecomment-1325100914).

This is not part of SQL standard, no need to update `sql_features`.

Referenced:  https://github.com/antlr/grammars-v4/tree/master/sql/postgresql

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by jeeminso** [comment]
Question: The current implementation unquotes the dollar quoting in the very beginning while parsing the statement. Are there any scenarios that we need the original form of the string again?

File: docs/sql/general/lexical-structure.rst [hunk 1]
```diff
@@ -28,6 +28,22 @@ String literals are defined as an arbitrary sequence of characters that are
 delimited with single quotes ``'`` as defined in ANSI SQL, for example
 ``'This is a string'``.
 
+In addition, CrateDB supports dollar quoted strings to help avoid escaping
+single quotes within single quoted strings.
+For example, ``'I''m a string'`` can be re-written as
+``$TAG$I'm a string$TAG$``, where the matching pair of ``TAGs`` can be zero or
```
Comment (matriv) [inline docs/sql/general/lexical-structure.rst:34]:
```suggestion
``$TAG$I'm a string$TAG$``, where the matching pair of ``TAG`` can be zero or
```
why did you use plural here?

File: libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java [hunk 2]
```diff
@@ -185,7 +186,7 @@ public void testEmptyExpression() {
         assertThatThrownBy(
             () -> SqlParser.createExpression(""))
             .isExactlyInstanceOf(ParsingException.class)
-            .hasMessageStartingWith("line 1:1: mismatched input '<EOF>' expecting {'(', '[', '[]', '{',");
+            .hasMessageStartingWith("line 1:1: mismatched input '<EOF>' expecting {");
```
Comment (matriv) [inline libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:189]: Why has this changed regarding the `expecting` part?

File: libs/sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java [hunk 3]
```diff
@@ -349,7 +349,7 @@ public void test_decommission_node_id_is_missing() {
         assertThatThrownBy(
             () -> printStatement("ALTER CLUSTER DECOMMISSION'"))
             .isExactlyInstanceOf(ParsingException.class)
-            .hasMessageStartingWith("line 1:27: mismatched input ''' expecting {'(', '[', '[]', '{', '$', '?', ");
+            .hasMessageStartingWith("line 1:27: mismatched input ''' expecting {");
```
Comment (matriv) [inline libs/sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java:352]: Same here and below.

File: libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java [hunk 4]
```diff
@@ -511,6 +512,49 @@ public void test_special_char_data_type() {
         assertThat(cast.getType().name()).isEqualTo("\"char\"");
     }
 
+    @Test
+    public void test_dollar_quoted_strings_with_valid_tags() {
+        assertThat(((StringLiteral) SqlParser.createExpression("$$$$")).getValue()).isEqualTo("");
```
Comment (matriv) [inline libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:517]: minor, no need to change, you can use `isEmpty()` assertion.

File: libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java [hunk 5]
```diff
@@ -511,6 +512,49 @@ public void test_special_char_data_type() {
...
+            .isInstanceOf(ParsingException.class)
+            .hasMessageStartingWith("line 1:5: extraneous input '$$' expecting");
+        assertThatThrownBy(
+            () -> SqlParser.createExpression("$1$$1$")) // $1$ is an invalid tag
+            .isInstanceOf(ParsingException.class)
+            .hasMessageStartingWith("line 1:3: mismatched input '$$' expecting");
+        assertThatThrownBy(
+            () -> SqlParser.createExpression("$a$$b$"))
+            .isInstanceOf(ParsingException.class)
+            .hasMessageStartingWith("line 1:7: extraneous input '<EOF>' expecting {DOLLAR_QUOTED_STRING_BODY, END_DOLLAR_QUOTED_STRING}");
```
Comment (matriv) [inline libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:555]: Imho, this is not nice for the user, can we have the old behaviour where we list the exact tokens and not the lexer names?

Update, now I get why the change in my other comments.

**Review by matriv — COMMENTED** [review]
Thanks a lot for your work here!
- Imho, using internal lexer names in user facing error messages is not nice
- Can we have some test with function definition?

> Question: The current implementation unquotes the dollar quoting in the very
> beginning while parsing the statement. Are there any scenarios that we need the 
> original form of the string again?

I don't think this is an issue, as long as the error message show the correct column number where the error is detected.

File: docs/sql/general/lexical-structure.rst [hunk 6]
```diff
@@ -28,6 +28,22 @@ String literals are defined as an arbitrary sequence of characters that are
...
+::
+
+    cr> select 'I''m a string' = $tag1$I'm a string$tag1$;
+    +------+
+    | true |
+    +------+
+    | TRUE |
+    +------+
+    SELECT 1 row in set (... sec)
+
```
Comment (matriv) [inline docs/sql/general/lexical-structure.rst:46]: Maybe add a note that we don't support nesting tags: `$a$......$b$......$b$.....$a$`

File: docs/sql/general/lexical-structure.rst (hunk 1 above)
Comment (jeeminso) [reply docs/sql/general/lexical-structure.rst:34]: fixed, Thanks

File: libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java (hunk 2 above)
Comment (jeeminso) [reply libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:189]: it is changed due to the commits and also `mismatched input...` seemed what was important here.

File: libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java (hunk 5 above)
Comment (jeeminso) [reply libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:555]: To my understanding, the behaviour did not change. Before, `mismatched input '<EOF>' expecting {'(', '[', ...` message looked like it contained the exact tokens because '(`, `[`, etc were implicitly defined.

Comment (matriv) [reply libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:555]: Can you take a look here: https://www.antlr2.org/doc/err.html#Modifying%20Default%20Error%20Messages%20With%20Paraphrases ?

Comment (matriv) [reply libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:555]: hm, this is for antlr2, but maybe something like that exists for antlr4

Comment (jeeminso) [reply libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:555]: Thank you, I will have a look at it now. Just to clarify, instead of `expecting {DOLLAR_QUOTED_STRING_BODY` would you expect to see `expecting { ~ '$'+ }` ?

Comment (matriv) [reply libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:555]: The best I could find is this: https://github.com/crate/crate/commit/9ed5b86cd9acec8fd301e998b050f8c523863fe2
I've also replaced some left over literals in the parser grammar with token names from lexer.

Comment (matriv) [reply libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java:555]: > Thank you, I will have a look at it now. Just to clarify, instead of `expecting {DOLLAR_QUOTED_STRING_BODY` would you expect to see `expecting { ~ '$'+ }` ?

Something like that yes, or check what I've used in my commit ^^.

Comment (jeeminso) [reply libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java]: Thanks, it looks good. I pushed it

File: docs/sql/general/lexical-structure.rst [hunk 7]
```diff
@@ -28,6 +28,26 @@ String literals are defined as an arbitrary sequence of characters that are
 delimited with single quotes ``'`` as defined in ANSI SQL, for example
 ``'This is a string'``.
 
+In addition, CrateDB supports dollar quoted strings to help avoid escaping
+single quotes within single quoted strings.
+For example, ``'I''m a string'`` can be re-written as
+``$TAG$I'm a string$TAG$``, where the matching pair of ``TAG`` can be zero or
```
Comment (matriv) [inline docs/sql/general/lexical-structure.rst:34]: I don't get this phrase: `can be zero or more characters`?

File: libs/sql-parser/src/main/java/io/crate/sql/parser/ExceptionErrorStrategy.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        beginErrorCondition(recognizer);
+
+        Token t = recognizer.getCurrentToken();
+        String tokenName = getTokenErrorDisplay(t);
+        IntervalSet expecting = getExpectedTokens(recognizer);
+        String msg = "extraneous input " + tokenName + " expecting " +
+                     expecting.toString(recognizer.getVocabulary());
+        msg = msg.replace("DOLLAR_QUOTED_STRING_BODY", "<dollar quoted string>")
```
Comment (matriv) [inline libs/sql-parser/src/main/java/io/crate/sql/parser/ExceptionErrorStrategy.java:50]: WDYT about: `$<tag>$<quoted string>$<tag>$`? instead of using wording?

File: libs/sql-parser/src/main/java/io/crate/sql/parser/ExceptionErrorStrategy.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            .replace("END_DOLLAR_QUOTED_STRING", "<end of dollar tag>");
```
Comment (matriv) [inline libs/sql-parser/src/main/java/io/crate/sql/parser/ExceptionErrorStrategy.java:51]: same here: `closing $<tag>$ for quoted string`

**Review by matriv — COMMENTED** [review]
I left a comment about docs, and a different suggestion about the error message, but maybe it's too much and it's fine as is already, let me know your opinion as well.

File: docs/sql/general/lexical-structure.rst (hunk 7 above)
Comment (jeeminso) [reply docs/sql/general/lexical-structure.rst:34]: Thanks, I appended `in length`, hope it is clearer.

**Review by matriv — APPROVED** [review]
Thank you!

**Comment by jeeminso** [comment]
retest this please

unrelated: `Could not resolve com.github.jk1:gradle-license-report:2.1`

**Comment by jeeminso** [comment]
retest this please
unrelated : `testTransientErrorsDuringRecoveryAreRetried`
